### PR TITLE
feat: support Merge-Into V1

### PR DESCRIPTION
**Pull request by JackTan25**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary
[snowflake merge-into](https://docs.snowflake.com/en/sql-reference/sql/merge)
[postgres merge-into](https://www.postgresql.org/docs/current/sql-merge.html)
Summary about this PR
merge into semantic: 
1. for one row in target_table, if there is over one row matched with it from the source, we need to throw an exception. (delta_lake semantic)
2. when there are no matched clauses for the merge command and when there is nothing matched for the merge command even if there are matched clauses, use merge_insert_only (delta_lake implementation)
3. delta_lake support table, but we need to do more like streaming,streaming_v2,values, this is hard to build plan.In delta lake, The source can be anything that you can turn into a DataFrame (when using the Scala or Python API) or anything that you can turn into a SQL query (when using the SQL API). It could even be the result of joining a 100 tables. 

[delta-lake implementation survey](https://datafuselabs.feishu.cn/docx/DAmDdVtypoH00IxzwnwcOq38nnn)
- Closes #issue
#10174

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/12350)

image
a logic implementation for V1 . V1 is not ready for users, it's just a testing version. we will do more optimizations based on it.
Some optimizations in the future:

[tracking](https://github.com/datafuselabs/databend/issues/12595)

### CODE REVIEW

**Comment by JackTan25** [comment]
I need to transform the merge into statement as a select ,build the insert_source as a TableReference firstly.

**Comment by JackTan25** [comment]
test：1. test type mismatch 2. columns sums mismatch 3. multi cluases 4. computed columns test

File: src/query/settings/src/settings_default.rs [hunk 1]
```diff
@@ -326,6 +326,12 @@ impl DefaultSettings {
                     possible_values: None,
                     display_in_show_settings: true,
                 }),
+                ("enable_unstable_merge_into", DefaultSettingValue {
```
Comment (BohuTANG) [inline src/query/settings/src/settings_default.rs:329]: `enable_experimental_merge_into` is better.

File: src/query/ast/src/ast/statements/merge_into.rs [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+pub enum MergeSource {
+    StreamingV2 {
+        settings: BTreeMap<String, String>,
+        on_error_mode: Option<String>,
+        start: usize,
+    },
+
+    Select {
+        query: Box<Query>,
+    },
```
[the remark this answers is not in the document]
Comment (JackTan25) [reply src/query/ast/src/ast/statements/merge_into.rs:211]: thanks I will add values, streaming source, table for next pr.

[the remark this answers is not in the document]
Comment (b41sh) [reply src/query/ast/src/ast/statements/merge_into.rs:211]: ok

File: src/query/storages/fuse/src/operations/merge_into/mutator/merge_into_split_mutator.rs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        let mut matched_block: Option<DataBlock> = None;
+        let mut not_matched_block: Option<DataBlock> = None;
+
+        for row_idx in 0..block.num_rows() {
+            let single_row = block.slice(row_idx..row_idx + 1);
+            if self.is_matched(row_id_column, row_idx)? {
+                if matched_block.is_none() {
+                    matched_block = Some(single_row);
+                } else {
+                    matched_block = Some(DataBlock::concat(&[matched_block.unwrap(), single_row])?);
```
[the remark this answers is not in the document]
Comment (JackTan25) [reply src/query/storages/fuse/src/operations/merge_into/mutator/merge_into_split_mutator.rs:55]: If I get the bitmap from NullableColumn, the Null will be zero, non-null will be set, right? cc @b41sh

File: src/common/exception/src/exception_code.rs [hunk 4]
```diff
@@ -185,6 +185,7 @@ build_exceptions! {
     BackgroundJobAlreadyExists(1501),
     UnknownBackgroundJob(1502),
 
+    InValidRowIdIndex(1503),
```
Comment (andylokandy) [inline src/common/exception/src/exception_code.rs:188]: do you mean this?

```suggestion
    InvalidRowIdIndex(1503),
```

File: src/query/expression/src/kernels/filter.rs [hunk 5]
```diff
@@ -70,6 +72,50 @@ impl DataBlock {
         }
     }
 
+    pub fn split_filter_with_bitmap(self, bitmap: &Bitmap) -> Result<(DataBlock, DataBlock)> {
```
Comment (andylokandy) [inline src/query/expression/src/kernels/filter.rs:75]: Is it benefit in performance comparing to:
```rust
let true_rows = block.filter_with_bitmap(filter)?;
let false_rows = block.filter_with_bitmap(!filter)?;
```

File: src/query/ast/src/parser/statement.rs [hunk 6]
```diff
@@ -151,6 +151,37 @@ pub fn statement(i: Input) -> IResult<StatementMsg> {
...
+            _,
+            (catalog, database, table),
+            alias_target,
+            _,
+            source,
+            _,
+            join_expr,
+            merge_options,
+        )| {
+            // we can get the target_table_reference in later, it doesn't matter.
```
Comment (andylokandy) [inline src/query/ast/src/parser/statement.rs:171]: what does this comment mean?

File: src/query/expression/src/kernels/filter.rs (hunk 5 above)
Comment (JackTan25) [reply src/query/expression/src/kernels/filter.rs:75]: avoid data block clone, here is self, not &self.

File: src/query/ast/src/parser/statement.rs (hunk 6 above)
Comment (JackTan25) [reply src/query/ast/src/parser/statement.rs:171]: The TableReference Will generate in bind_merge_into. Let me remove this.

File: src/common/exception/src/exception_code.rs (hunk 4 above)
Comment (JackTan25) [reply src/common/exception/src/exception_code.rs:188]: thanks.

File: src/query/expression/src/kernels/filter.rs (hunk 5 above)
Comment (andylokandy) [reply src/query/expression/src/kernels/filter.rs:75]: datablock clone is merely `Arc::clone`

File: src/query/storages/fuse/src/operations/merge_into/mutator/merge_into_split_mutator.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (b41sh) [reply src/query/storages/fuse/src/operations/merge_into/mutator/merge_into_split_mutator.rs:55]: yes

**Comment by dantengsky** [comment]
[linux / sqllogic_standalone_base_parquet](https://github.com/datafuselabs/databend/actions/runs/6000760718/job/16292350787?pr=12350#logs) failed

~~~
0: statement failed: mysql client error: Server error: `ERROR HY000 (1105): InvalidRowIdIndex. Code: 1503, Text = row id column should be a column, but it's a scalar.'
[SQL] merge into t1 using (select * from t2 as t2) on t1.a = t2.a  when matched then update set t1.c = t2.c  when not matched then insert (a,b,c) values(t2.a,t2.b,t2.c);
at tests/sqllogictests/suites/base/09_fuse_engine/09_0026_merge_into:73
~~~


https://github.com/datafuselabs/databend/actions/runs/6000760718/job/16292350787?pr=12350#step:4:838

**Comment by JackTan25** [comment]
There are some optimizations advised by @b41sh , I will do it firstly. For now, I use some concat in matched and not matched, there will be many memory copy, and I will use update expr (just use only the one and the same block, no need to split). Thanks

**Review by dantengsky — APPROVED** [review]
:+1: LGTM

**Review by b41sh — APPROVED** [review]

**Pull request merged**